feature(azure devops): support multiple organisations

sanderaernouts opened the pull request:
## Hey, I just made a Pull Request!

I Added an `AzureDevOpsCredentialProvider`, similar to the GitHub and AWS integrations, that can return a token based on the provided URL. This PR resolves #10431 by adding support for `organisations` to the Azure integration config.

The updated Azure integration configuration looks like this:
```yaml
integrations:
  azure:
  - host: my.devops.server
    credentials:
    - token: my-pat
  - host: dev.azure.com
    credentials:
    - clientId: <id>
      clientSecret: <secret>
      tenantId: <tenant>
      organisations:
      - my-org
      - my-other-org
      - yet-another-org
    - token: my-org-pat
      organisations:
      - my-other-org
```

The current `token` and `credential` fields in the Azure integration config are deprecated in favour of the `credentials` field

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] ~Screenshots attached (for UI changes)~
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

sanderaernouts commented:
@awanlin, no, I won't be touching that part. I typed the YAML config example from memory and forgot that the Azure integration config is an array of configs. To clarify, I plan to replace the `token` and `credential` fields with a single `credentials` field. This field is an array of token, service principal or managed identity credentials. The credentials will have an optional `organisations` field where you can specify which Azure DevOps organisation the credential applies.
```yaml
integrations:
  azure:
    - host: dev.azure.com
      credentials: 
      - organisations:
        - my-org
        token: my-pat 
      - organisations:
        - my-other-org
        token: my-other-pat
    - host: some.devops.server
      credentials: 
      - token: my-pat 
```

I have updated the PR description 👍

awanlin commented:
Awesome, thanks for the follow up @sanderaernouts. Another question popup though: why the extra `organizations`?

Could this:

```yaml
integrations:
  azure:
    - host: dev.azure.com
      credentials: 
      - organisations:
        - my-org
        token: my-pat 
      - organisations:
        - my-other-org
        token: my-other-pat
    - host: some.devops.server
      credentials: 
      - token: my-pat 
```

Be something like this instead:

```yaml
integrations:
  azure:
    - host: dev.azure.com
      credentials: 
        - organization: my-org
          token: my-pat 
        - organization: my-other-org
          token: my-other-pat
    - host: some.devops.server
      credentials:
        - organization: my-org
          token: my-pat
```

With Azure DevOps Server the concept of Team Collections is more or less the same as Organizations in Azure DevOps Services. I think going this way seeing as we will have a breaking change makes this more consistent

sanderaernouts commented:
@awanlin I added an array of organizations to support reusing the same service principal or managed identity for multiple organizations. A single service principal or managed identity could be used as long as the same Azure AD tenant backs the Azure DevOps organizations.

However, I'm also okay with `organization`; you can still duplicate the service principal or managed identity.

sanderaernouts commented:
@awanlin I updated the PR description and will use similar examples when updating the docs. I have also added an example for the Azure DevOps server. Meanwhile, I made good progress on the credential provider yesterday, so there already is some stuff to review if you want to.

Rugvip commented on packages/backend-common/src/reading/AzureUrlReader.ts:
```diff
@@ -47,14 +48,19 @@ import { ReadUrlResponseFactory } from './ReadUrlResponseFactory';
   constructor(
+    private readonly provider: AzureDevOpsCredentialsProvider,
     private readonly integration: AzureIntegration,
     private readonly deps: { treeResponseFactory: ReadTreeResponseFactory },
```
Let's pass the `provider` through `deps` instead. I'm feeling it's best to call it `credentialsProvider` too, to keep that clearer in the implementation

sanderaernouts replied:
@Rugvip, what do you mean by this?
> let's pass the provider through deps instead

Should I modify the `ReaderFactory` to something like this?
```typescript
static factory: ReaderFactory = ({ config, treeResponseFactory, credentialsProvider }) => { ... }
```

From what I can tell, that, in turn, would mean that the `UrlReadersOptions` also needs an `azureCredentialsProvider` property; then the `azureCredentialsProvider` is provided as part of `makeCreateEnv`?
```typescript
function makeCreateEnv(config: Config) {
  const root = getRootLogger();
  const reader = UrlReaders.default({ logger: root, config, azureCredendialsProvider });
  //...
```

Since the provider caches tokens, we could go with `getAzureCredentialsProvider` and `seAzureCredentialsProvider` functions similar to `getRootLogger()` and `setRootLogger()`.

Or do you mean that the `AzureCredentialsProvider` should become part of the [core services](https://backstage.io/docs/backend-system/architecture/services/#core-services)? 

I checked how this is done for the `DefaultGithubCredentialsProvider`, and that provider also instantiated inside the `factory` method:
```typescript
export class GithubUrlReader implements UrlReader {
  static factory: ReaderFactory = ({ config, treeResponseFactory }) => {
    const integrations = ScmIntegrations.fromConfig(config);
    const credentialsProvider =
      DefaultGithubCredentialsProvider.fromIntegrations(integrations);
    return integrations.github.list().map(integration => {
      const reader = new GithubUrlReader(integration, {
        treeResponseFactory,
        credentialsProvider,
      });
      //...
```
 https://github.com/backstage/backstage/blob/698b0d749b9c5885c39866a0569663f799b0892a/packages/backend-common/src/reading/GithubUrlReader.ts#LL64C5-L65C71

sanderaernouts replied:
I didn't read your comment carefully enough 😓. I modified the code to pass the `credentialsProviders` through the `deps` object.

Rugvip commented on packages/backend-common/src/reading/AzureUrlReader.ts:
```diff
@@ -113,7 +119,7 @@ export class AzureUrlReader implements UrlReader {
     const commitsAzureResponse = await fetch(
       getAzureCommitsUrl(url),
-      await getAzureRequestOptions(this.integration.config),
+      await this.provider.getCredentials({ url: url }),
```
Thinking it's best to explicitly pass just `headers` here, just in case

sanderaernouts replied:
Done 👍

Rugvip commented on packages/integration/src/azure/config.ts:
```diff
@@ -33,25 +33,26 @@ export type AzureIntegrationConfig = {
+   * If no credentials are specified at all, anonymous access is used.
    */
-  token?: string;
+  credentials?: AzureIntegrationCredentialConfig[];
```
We can't remove the `token` at this point, it'd be an immediate breaking change without deprecation period. If we feel it makes sense to remove in the future then we could deprecate it. I'm not entirely convinced we should do it, although in particular if we want to nudge people towards other options then I think it makes sense.

For now I suggest we generate a default credentials entry using the token if configured. It might also make sense to reject config that lists both a top-level token and credentials

sanderaernouts replied:
Done, I marked both `token` and `credential` as `@deprecated`. I did the scaffolder and catalog plugins for Azure to use the new `credentials` format.

sanderaernouts replied:
I also moved `getAzureRequestOptions` to `deprecated.ts` after removing all references to it from the code base. That way plugins that use this function will not break instantly.

Rugvip commented on packages/integration/src/azure/config.ts:
```diff
@@ -33,25 +33,26 @@ export type AzureIntegrationConfig = {
  * @public
  */
-export type ClientSecret = {
+export type ClientSecretCredential = AzureIntegrationCredentialConfigBase & {
```
I like this rename and I think we could take it one step further, `AzureClientSecretCredential`. After the release next week this will be a breaking change. What do you think about moving these renames to a separate PR that we can ship a bit faster?

sanderaernouts replied:
Sure thing, see #18273. I'll rebase this branch on those changes

Rugvip commented on packages/integration/src/azure/config.ts:
```diff
@@ -71,40 +72,73 @@ export type ClientSecret = {
 
   return (
-    Object.keys(credential).length === 3 &&
+    getConfigKeys(credential).length === 3 &&
```
Do we really need to count keys here? 😅 is it not enough to check for the presence of the 3 props?

sanderaernouts replied:
Probably not. The reason I did this was to prevent configs like this:
```yaml
host: dev.azure.com
credentials: 
  - token: token
    clientId: id
    clientSecret: secret
    tenantId: tenant
```` 
This config would match `AzureClientSecretCredential`, `AzureManagedIdentityCredential`, and `PersonalAccessTokenCredential`. But I can probably prevent this without counting keys.

sanderaernouts replied:
Fixed, I refactored the type predicates to check the specific properties instead.

Rugvip reviewed (changes requested):
Nice, great to get this added for azure too! 👍 

Couple of initial comments. In particular I think it's best if we split out renames of existing systems in a new PR to avoid breaking changes, because I don't think this overall is something that we can ship straight into a main-line release.

sanderaernouts commented:
@awanlin, I updated the docs as well. Should we restrict personal access token (PAT) credentials to zero or one organization in the integration config? For Azure DevOps, the PAT is organization-specific, so filling out more than one organization makes no sense.

sanderaernouts commented:
@Rugvip I deprecated the fields and functions instead of removing them. I don't think the changes are breaking anymore. I'm unsure about the `AzureUrlReader`, though. I changed the constructor, but it seems a static factory (`create`) is used to construct it.

awanlin commented on docs/integrations/azure/locations.md:
```diff
@@ -52,13 +76,18 @@ integrations:
+
+The `credentials` element is a structure with these elements:
+
+- `organizations`: (optional): A list of organizations for which this credential should be used. Only supported for Azure DevOps (cloud). If not specified the credential will be used for all organizations for which no other credential is configured.
```
Why would this be cloud only? This should work with Server as well but only with a PAT or am I missing something here? Would be really disappointing if that is the case

awanlin replied:
If this does work with Server we should mention that Organizations is synonymous with Team Project Collections

sanderaernouts replied:
I cannot confirm this because I do not have access to an Azure DevOps Server. But from what I remember, a personal access token is server-wide on-premise, depending on the user's permissions. You can already add multiple Azure DevOps servers because they have different host names, so I don't see a need for separate credentials per collection.

sanderaernouts replied:
I did not implement support for the team project collection URLs in the `DefaultAzureDevOpsCredentialsProvider` but that would be simple to add if it does make sense.

awanlin commented:
> @awanlin, I updated the docs as well. Should we restrict personal access token (PAT) credentials to zero or one organization in the integration config? For Azure DevOps, the PAT is organization-specific, so filling out more than one organization makes no sense.

Yeah, that makes sense to me 👍

sanderaernouts commented:
> > @awanlin, I updated the docs as well. Should we restrict personal access token (PAT) credentials to zero or one organization in the integration config? For Azure DevOps, the PAT is organization-specific, so filling out more than one organization makes no sense.
> 
> Yeah, that makes sense to me 👍

Done ✅

sanderaernouts commented:
I ran some tests locally with a client secret credential for one AAD-backed Azure DevOps organization and a PAT for my personal Azure DevOps organization. I was able to read `catalog-info.yaml` files from both organizations 🥳.

sanderaernouts commented:
Updated the PR description since it is no longer a work in progress

sanderaernouts commented:
@awanlin @Rugvip can you have another look? 👀

sanderaernouts commented:
👍 I'm on the Backstage discord server, so if you run into anything while testing, you can ping me there as well. My Discord nickname is the same as my GitHub handle

awanlin commented:
OK, did some testing with the environments I had:

- Multiple Team Collections on Azure DevOps Server 2020 worked
- Multiple Organizations on Azure DevOps Services work
- A combination of Team Collections on Azure DevOps Server 2020 and Organizations on Azure DevOps Services works

Awesome work @sanderaernouts, once this gets merged in I'll work on updating the Azure DevOps plugin to support this 🚀

sanderaernouts commented:
@Rugvip can you have another look? Had some questions based on review

awanlin commented:
@sanderaernouts just a heads up that @Rugvip is on vacation for a few weeks now, sorry bad timing I know

Oh, but in the mean time if you rebase your E2E tests will start passing 👍

sanderaernouts commented:
@awanlin done ✅ . Is there anyone else who can have a look at this PR instead of @Rugvip? I was hoping to get this out with `1.16.0` 🚀.

From his review, there is one unresolved comment, https://github.com/backstage/backstage/pull/18213#discussion_r1230223085, but I need some guidance on how to resolve that. So if anyone can help me resolve that, that would be great.

sanderaernouts commented:
It turns out I did not read [@Rugvip's comment](https://github.com/backstage/backstage/pull/18213#discussion_r1230223085) carefully enough. I refactored the `AzureUrlReader` class and have resolved the comment 👍 .

awanlin commented:
> It turns out I did not read [@Rugvip's comment](https://github.com/backstage/backstage/pull/18213#discussion_r1230223085) carefully enough. I refactored the `AzureUrlReader` class and have resolved the comment 👍 .

The soonest I'll have time to retest this is On Friday July 7th

Rugvip commented on packages/integration/src/azure/CachedAzureDevOpsCredentialsProvider.ts:
```diff
@@ -0,0 +1,116 @@
+    if (
+      this.cached === undefined ||
+      (this.cached.expiresAt !== undefined &&
+        DateTime.local() > this.cached.expiresAt)
```
I'd expect `DateTime.utc()` to be the way to go here with the `DateTime.fromSeconds` call below also using an explicit tz, just doesn't feel quite right to use the local tz, especially with things like daylight saving. Or tbh I feel it could be moved over to `Date.now()` + math too

sanderaernouts replied:
Went with `Date.now()`

Rugvip commented on packages/integration/src/azure/CachedAzureDevOpsCredentialsProvider.ts:
```diff
@@ -0,0 +1,116 @@
+
+        this.cached = {
+          headers: {
+            Authorization: `Basic ${buffer.toString('base64')}`,
```
Could go for ``btoa(`:${this.credential}`)`` here, since that's now available in node

sanderaernouts replied:
Fixed 👍

Rugvip commented on packages/integration/src/azure/CachedAzureDevOpsCredentialsProvider.ts:
```diff
@@ -0,0 +1,116 @@
+          this.azureDevOpsScope,
+        );
+
+        if (accessToken === null) {
```
Feeling that this could be `!accessToken` just in case, explicit `null` check feels a bit strict here and a bit tricky to debug if empty return is switched to `undefined`.

sanderaernouts replied:
Fixed 👍

Rugvip reviewed (changes requested):
Nice, thank you! 😁 

Bunch of comments but overall it looks solid 👍

sanderaernouts commented on packages/backend-common/api-report.md:
```diff
@@ -91,6 +92,7 @@ export class AwsS3UrlReader implements UrlReader {
 // @public
 export class AzureUrlReader implements UrlReader {
   constructor(
+    provider: AzureDevOpsCredentialsProvider,
```
Done ✅

sanderaernouts replied:
Done ✅

sanderaernouts commented on packages/backend-common/src/reading/AzureUrlReader.ts:
```diff
@@ -47,16 +48,24 @@ import { ReadUrlResponseFactory } from './ReadUrlResponseFactory';
 export class AzureUrlReader implements UrlReader {
   static factory: ReaderFactory = ({ config, treeResponseFactory }) => {
     const integrations = ScmIntegrations.fromConfig(config);
+    const credentialProvider =
+      DefaultAzureDevOpsCredentialProvider.fromIntegrations(integrations);
```
Done ✅

sanderaernouts commented on packages/integration/src/azure/config.ts:
```diff
@@ -117,45 +199,117 @@ export function readAzureIntegrationConfig(
-    : undefined;
+      },
+    ];
+    credentials = credentials?.concat(mapped) ?? mapped;
```
You are right; the `credential` field should go away. The only reason I did not remove it already is to have a deprecation period instead of a breaking change. A credentials with no `organizations` will be used as a fallback credential so it makes no sense to support both `credential` and `credentials`

sanderaernouts commented on packages/integration/src/azure/config.ts:
```diff
@@ -68,44 +91,103 @@ export type AzureClientSecretCredential = {
+ */
+export function isAzureClientSecretCredential(
+  credential: AzureDevOpsCredentialLike,
+): credential is AzureClientSecretCredential {
```
This took me a while, but I got it working in the end :)

sanderaernouts replied:
@Rugvip I refactored `AzureDevOpsCredential` into a discriminated type union, but I wanted to keep the `kind` out of the configuration since that would make the configuration more verbose. But that means I must map an object similar to a credential (`AzureDevOpsCredentialLike`) to an object that satisfies the `AzureDevOpsCredential` type union. Even if I put `kind` in the config, I'd still have to validate that all required properties are present. 

So I added a mapping of credential kinds and fields and the asAzureDevOpsCredential` method in `config.ts` to perform the validation and mapping. 

I'm still trying to decide whether a more explicit and verbose implementation, for example, using specific functions to validate and map each credential kind, is better. I'm also trying to decide if it's better to put `kind` in the config. 

I would appreciate your opinion on this!

sanderaernouts commented on plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/azure.ts:
```diff
@@ -135,22 +142,23 @@ export function createPublishAzureAction(options: {
+      if (credentials === undefined && ctx.input.token === undefined) {
         throw new InputError(
-          `No matching integration configuration for host ${host}, please check your integrations config`,
+          `No credentials provided for Azure Integration ${host}`,
```
Went with `url` since it includes both the host and the organization.

sanderaernouts commented on packages/integration/src/azure/DefaultAzureDevOpsCredentialProvider.ts:
```diff
@@ -0,0 +1,85 @@
+          credential.organizations === undefined ||
+          credential.organizations.length === 0
+        ) {
+          if (acc.get(`${integration.config.host}`) === undefined) {
```
Fixed ✅

Rugvip commented on .changeset/six-wolves-admire.md:
```diff
@@ -0,0 +1,66 @@
+---
+'@backstage/integration': minor
+---
+
```
Bit too much text for a changeset here tbh, best just explain what's changed quick and then you can put a reference to the documentation to explain how it all works.

sanderaernouts replied:
I shortened the changeset 👍.

Rugvip commented on packages/integration/src/azure/DefaultAzureDevOpsCredentialsProvider.ts:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    if (provider === undefined) {
+      provider = this.providers.get(`${url.host}`);
```
Another one 😅

sanderaernouts replied:
At least I'm consistent 🤣.

sanderaernouts replied:
Fixed it 👍

sanderaernouts commented:
@awanlin, do you want to retest this based on the changes I made following the review from @Rugvip?

awanlin commented:
@sanderaernouts sure, that would be wise to do. I can only do this on Friday. I'm fine with this getting merged before then but only after the release comes out on Tuesday. 👍

sanderaernouts commented:
@awanlin since it's not going to be in the release today we might as well be on the safe side and retest it 👍. FYI, I have an open review comment in #18507 that might change how the Azure credentials are read from the config.

awanlin commented:
Retested just now and it's all still working as expected. The test cases I used were:

- Multiple Team Collections on Azure DevOps Server 2020 worked
- Multiple Organizations on Azure DevOps Services work
- A combination of Team Collections on Azure DevOps Server 2020 and Organizations on Azure DevOps Services works

sanderaernouts commented:
Thanks @awanlin! @Rugvip can you have another look?

sanderaernouts commented:
Tagging the other maintainers: @freben, @benjdlambert, @jhaals. This PR has been open for almost 2 months now, I know it's holiday season, but I'd love to get into `1.1.17`.

MarcBruins commented:
Any progress on this? We would very much like to have this supported.

afscrome commented on packages/integration/src/azure/CachedAzureDevOpsCredentialsProvider.ts:
```diff
@@ -0,0 +1,111 @@
+        // eslint-disable-next-line no-case-declarations
+        const _exhaustiveCheck: never = credential;
+
+        throw new Error(
```
Can we use `DefaultAzureCredential` as the fallback here.  This would allow the behaviour to "just work" when other components are configured with this approach.

sanderaernouts replied:
I refactored the `DefaultAzureDevOpsCredentialsProvider` to add a `CachedAzureDevOpsCredentialProvider` with a  `DefaultAzureCredential` when no credentials are specified for a `dev.azure.com` integration. For Azure DevOps server, the fallback remains to use no authentication.

afscrome commented on .changeset/six-wolves-admire.md:
```diff
@@ -0,0 +1,22 @@
+        - organizations:
+            - my-org
+            - my-other-org
+          clientId: ${CLIENT_ID}
+          clientSecret: ${CLIENT_SECRET}
+          tenantId: ${TENANT_ID}
```
Better match typical environment variables used by the Azure Identity libraries.
```suggestion
          clientId: ${AZURE_CLIENT_ID}
          clientSecret: ${AZURE_CLIENT_SECRET}
          tenantId: ${AZURE_TENANT_ID}
```
https://learn.microsoft.com/en-us/dotnet/api/azure.identity.environmentcredential?view=azure-dotnet

sanderaernouts replied:
Renamed the variables, also renamed them in the documentation

afscrome commented on docs/integrations/azure/locations.md:
```diff
@@ -52,13 +76,19 @@ integrations:
-> - `token` and `credential` are mutually exclusive.
 > - You cannot use a service principal or managed identity for Azure DevOps Server (on-premises) organizations
 > - You can only use a service principal or managed identity for Azure AD backed Azure DevOps organizations
+> - You cannot use more than one organization with a personal access token
```
Any reason for this?  When you create a PAT, you can generate a PAT for more than one organisation
![image](https://github.com/backstage/backstage/assets/289860/a7288681-ad49-42a8-8aad-963b5bb29960)

sanderaernouts replied:
I don't remember exactly why I added this restriction, but you are right it seems one or all organizations are allowed when creating a PAT, so I removed the restriction from the documentation and code.

afscrome commented on docs/integrations/azure/locations.md:
```diff
@@ -52,13 +76,19 @@ integrations:
+- `clientId`: The client ID of the service principal or managed identity (required for service principal and managed identities)
+- `clientSecret`: The client secret of the service principal (required for service principal)
+- `tenantId`: The tenant ID of the service principal (required for service principal)
+- `token`: The personal access token (required for personal access token)
```
If this is for a new `credentials` configuration array, is it worth renaming `token` to `personalAccessToken`.  `token` could be confused with `access token` used by the AAD auth methods?

sanderaernouts replied:
True, `token` is ambiguous, so I renamed it to `personalAccessToken`. It does seem redundant on the `PersonalAccessTokenCredential` type, but this type is mainly used as part of the `AzureDevOpsCredential` union type and in those cases `credential.personalAccessToken` is much more self-explanatory than `credential.token`

```typescript
export type PersonalAccessTokenCredential = AzureCredentialBase & {
  kind: 'PersonalAccessToken';
  personalAccessToken: string;
};
```

sanderaernouts commented:
@afscrome thanks for your review, I have processed your review comments 👍

sanderaernouts commented:
Refactored the `DefaultAzureDevOpsCredentialsProvider` a bit more so that it always uses a `DefaultAzureCredential` as the fallback for `dev.azure.com` unless a credential without `organizations` is specified. The `DefaultAzureDevOpsCredentialsProvider` always tries to find a credential for the specific organization before using a non organization specific credential.  So a `DefaultAzureCredential` is used unless:
- you specify a credential for the specific organization
- you specify a credential without `organizations`

sanderaernouts commented:
@awanlin tested this locally against two dev.azure.com organizations. Do you want to retest it for Azure DevOps server?

awanlin commented:
Ok, retested but I think we need to update the PR description, changeset, and the docs based on the latest changes. I needed to change `token` to `personalAccessToken` in my config, I'm fine with that but had to dig to figure that out. Also I needed to change the config structure as well and include new things I didn't need to add before. I was nice that you didn't need to list the org before.

I started with:

```yaml
azure:
    - host: azure-devop-server.company.com
      token: ${TOKEN}
    - host: dev.azure.com
      credentials:
      - token: ${ANOTHER_TOKEN}
        organisations:
        - some-org
 ```

and ended with:

```yaml
azure:
    - host: azure-devop-server.company.com
      credentials:
      - personalAccessToken: ${TOKEN}
        organisations:
        - server-collection
    - host: dev.azure.com
      credentials:
      - personalAccessToken: ${ANOTHER_TOKEN}
        organisations:
        - some-org
 ```

With the config above with changes to test the various scenarios it all worked

awanlin commented on .changeset/six-wolves-admire.md:
````diff
@@ -0,0 +1,22 @@
+Added `AzureDevOpsCredentialsProvider` to support multiple Azure DevOps organizations and **deprecated** `AzureIntegrationConfig.credential` in favour of `AzureIntegrationConfig.credentials`. You can now use specific credentials for different Azure DevOps organizations by specifying the `organizations` field on a credential:
+
+```yaml
+integrations:
````
I think we should include a PAT example as it's the most common setup right now

sanderaernouts replied:
Done 👍

awanlin commented on docs/integrations/azure/locations.md:
```diff
@@ -41,24 +41,53 @@ Using a personal access token (PAT):
     - host: dev.azure.com
-      token: ${AZURE_TOKEN}
+      credentials:
+        - personalAccessToken: ${PERSONAL_ACCESS_TOKEN}
```
Without an org this didn't work for me. That either needs to be added or maybe something needs to be fixed

sanderaernouts replied:
@awanlin I tried this myself, and for me, it does work. Can you elaborate a bit on what problem you are facing? Do you see any error messages?

Also, what does your config look like? If you, for example, have two `credentials` without an org, the first one listed will be used. In the case below, a `ManagedIdentityCredential` would be used because it's the first `credential` without an org:
```
- host: dev.azure.com
  credentials:
    - clientId: ....
    - personalAccessToken: ....
```

Even if this isn't the problem in your case, I can see how this could confuse the users. So I added two additional checks:
- reject the config when more than one credential without any `organizations` is specified
- reject the config when multiple credentials specify the same organization

p.s.: I want to get this into `1.1.17`, so can you poke one of the other maintainers to have a look as well? I mentioned @Rugvip a few times now, I mentioned the other @backstage/catalog-maintainers, and I dropped this PR in the `catalog` channel on Discord. However, I still need another review from a maintainer to get this merged.

awanlin replied:
There was no error it simply did not pull in the entity from my Azure DevOps Server instance but did for my Azure DevOps Cloud. I'll re-test today but as I mentioned I'm very worried about this going out in today's release. 

This is how my config was setup:

```yaml
azure:
    - host: azure-devop-server.company.com
      credentials:
      - personalAccessToken: ${TOKEN}
    - host: dev.azure.com
      credentials:
      - personalAccessToken: ${ANOTHER_TOKEN}
        organisations:
        - some-org
```

awanlin replied:
Ok, re-tested and I found the issue: when you create a PAT that is only for a specific org but don't include the org in the config like I did above it doesn't work. Before I wasn't getting any error but now I get a 401 which at least forced me to play around with this. The fix was that I had to create the PAT using the "For all accessible organizations" option. This same PAT (setup for 1 org) worked before in my prior testing .

awanlin replied:
When creating a PAT it always seems to default to one of the Orgs you have access to and not the "For all accessible organizations" option. At least that's what I get when I test with Server and Cloud, can you confirm? Hopefully I'm wrong. Based on that most adopters will have an Org specific PAT in place that they know is good, try to use it and run into this issue if they don't include the Org.

sanderaernouts replied:
Interesting 🤔 , I can't easily explain that based on the changes I made to support multi-org personal access tokens. I'm pretty sure that the PAT I'm using to test is in fact for a single organization, but I'll test both scenarios to see if I get the same error.

awanlin replied:
Ok, did more testing. My main focus is being able to pull from both Server and Cloud instances. In my testing that has been the setup I've tested the most. If I just use:

```yaml
integrations:
  azure:
    - host: sever.company.com
      credentials:
        - personalAccessToken: ${ORG_SPECIFIC_TOKEN}
``` 

That works fine and pulls in the entity I have setup. But as soon as I add in this setup I run into the issue I posted about earlier:

```
azure:
    - host: azure-devop-server.company.com
      credentials:
      - personalAccessToken: ${ORG_SPECIFIC_TOKEN}
    - host: dev.azure.com
      credentials:
      - personalAccessToken: ${ANOTHER_TOKEN}
        organisations:
        - some-org
```
For the above to work I need to add the org or use an org-wide PAT

sanderaernouts replied:
That is weird, the organization is only used for matching the URLs and when no credential for a specific organization is found then I try to find a credential without an organization.

I do not treat the PAT differently so I wonder why this doesn’t work for DevOps server when you use an org specific PAT. Tomorrow I have a day off, so I’ll look into it further on Thursday.

sanderaernouts replied:
@awanlin if I understand correctly it works when you specify only 1 azure integration but when you add a 2nd integration, for dev.azure.com in this case, it doesn’t work anymore and you have to specify the org for DevOps Server?

awanlin replied:
Yes, that's exactly what I'm running in to. I have not tested the reverse but can do that later today

sanderaernouts replied:
@awanlin I tried to replicate the scenario with your config, but it is impossible to specify an `organization` for an Azure DevOps server. You should get the following error:

So, modifying your configuration like this:

```diff
integrations:
  azure:
    - host: azure-devops-server.company.com
      credentials:
        - personalAccessToken: `${ORG_SPECIFIC_TOKEN}`
+         organizations:
+           - my-org
    - host: dev.azure.com
      credentials:
        - personalAccessToken: ${ANOTHER_TOKEN}
```

Results in this error message:
```text
Backend failed to start up Error: Invalid Azure integration config for azure-devops-server.company.com, you cannot specify organizations for hosts other than dev.azure.com
```

The backend only starts when I remove the `organizations` property again:
```diff
integrations:
  azure:
    - host: azure-devops-server.company.com
      credentials:
        - personalAccessToken: `${ORG_SPECIFIC_TOKEN}`
-         organizations:
-           - my-org
    - host: dev.azure.com
      credentials:
        - personalAccessToken: ${ANOTHER_TOKEN}
```

I don't have access to an Azure DevOps server, so I cannot test whether the `${ORG_SPECIFIC_TOKEN}` works for me, but the `dev.azure.com` integration works with an org-specific token without specifying the `organizations` field.

Did you mean you must specify the `organizations` field on the credential for `dev.azure.com` to make it work?
```diff
integrations:
  azure:
    - host: azure-devops-server.company.com
      credentials:
        - personalAccessToken: `${ORG_SPECIFIC_TOKEN}`
    - host: dev.azure.com
      credentials:
        - personalAccessToken: ${ANOTHER_ORG_SPECIFIC_TOKEN}
+         organizations:
+           - my-org
```

Also, I did notice that you used the British English form (organi**s**ations) in your example, whereas the config and code behind use the American English form (organi**z**ations). But I assume this is a typo in your GitHub comment instead of a typo in your configuration. It makes no sense that specifying the organi**s**ations field on a credential causes it to work.

awanlin replied:
@sanderaernouts just a heads up I can't re-test until the afternoon, I have meeting all morning. Also, would it make sense for us to move this to Discord? My user name is `Ahhhndre`. I'm also thinking that we should try get on a call as well that way you can get all the information you need

awanlin replied:
I know we are meeting shortly but just wanted to comment on the spelling, I have been using `organisations` in all my testing and never ran into any errors with this so that makes this all the more odd.

sanderaernouts replied:
@awanlin based on our meeting earlier today, I made the following changes:
- support `organizations` for Azure DevOps server (including servers prefixed with `/tfs`)
- updated the schema in `config.d.ts` to reflect the new Azure integration configuration

awanlin replied:
Awesome, will test now!

awanlin replied:
Testing is all good 🚀

Rugvip commented on .yarn/releases/yarn-3.2.3.cjs:
```diff
@@ -202,7 +202,7 @@ Please report this to https://github.com/szmarczak/http2-wrapper/`);p(m)}return}
 attempted value: ${s} 
 `+(o!==s?`result of cast: ${o}`:""))}return n}_cast(e,t){let i=e===void 0?e:this.transforms.reduce((n,s)=>s.call(this,n,e,this),e);return i===void 0&&(i=this.getDefault()),i}_validate(e,t={},i){let{sync:n,path:s,from:o=[],originalValue:a=e,strict:l=this.spec.strict,abortEarly:c=this.spec.abortEarly}=t,u=e;l||(u=this._cast(u,zs({assert:!1},t)));let g={value:u,path:s,options:t,originalValue:a,schem ...
-`),c=o.length;for(c>0&&o.includes("unstyle")&&(o=[...new Set(["unstyle",...o])].reverse());c-- >0;)a=t(r.styles[o[c]],a,l);return a},n=(s,o,a)=>{r.styles[s]=e({name:s,codes:o}),(r.keys[a]||(r.keys[a]=[])).push(s),Reflect.defineProperty(r,s,{configurable:!0,enumerable:!0,set(c){r.alias(s,c)},get(){let c=u=>i(u,c.stack);return Reflect.setPrototypeOf(c,r),c.stack=this.stack?this.stack.concat(s):[s], ...
+`),c=o.length;for(c>0&&o.includes("unstyle")&&(o=[...new Set(["unstyle",...o])].reverse());c-- >0;)a=t(r.styles[o[c]],a,l);return a},n=(s,o,a)=>{r.styles[s]=e({name:s,codes:o}),(r.keys[a]||(r.keys[a]=[])).push(s),Reflect.defineProperty(r,s,{configurable:!0,enumerable:!0,set(c){r.alias(s,c)},get(){let c=u=>i(u,c.stack);return Reflect.setPrototypeOf(c,r),c.stack=this.stack?this.stack.concat(s):[s], ...
```
The changes here need to be reverted

sanderaernouts replied:
Not sure how that happened 🤔. Removed it from the commit.

Rugvip commented on packages/integration/src/azure/CachedAzureDevOpsCredentialsProvider.ts:
```diff
@@ -0,0 +1,127 @@
+          new ManagedIdentityCredential(credential.clientId),
+        );
+      default:
+        // @ts-ignore
+        // eslint-disable-next-line no-case-declarations
+        const _exhaustiveCheck: never = credential;
```
Could you refactor this to note require a `@ts-ignore`? For example just making it an if-else chain, or I think `switch (credential.kind as string) {` might do it too.

sanderaernouts replied:
The `@ts-ignore` was there because `_exhaustiveCheck` was declared but never used. And  `_exhaustiveCheck: never` was there to generate a typescript error when you forget to cover a (new) case in the switch statement.

I was able to get rid of the `@ts-ignore` and as a bonus the `eslint-disable-next-line` by declaring a function for the exhaustive type check instead:
```diff
+ function exhaustiveCheck(_param: never) {}
....
- // @ts-ignore
- // eslint-disable-next-line no-case-declarations
- const _exhaustiveCheck: never = credential;
+ exhaustiveCheck(credential);
```

Rugvip reviewed (changes requested):
Nice, thank you! 🎉 

Just two nits. It would be nice to have some clarity on https://github.com/backstage/backstage/pull/18213#discussion_r1291717382 but I don't feel it's a requirement

There are a lot of PRs we want to get into 1.17 😅

Rugvip reviewed (approved):
✨

awanlin commented:
I'd just like to note my worries about this getting into the release today. I strongly feel that this needs to be included in at least one of the weekly releases.

Rugvip commented:
@awanlin will do!

awanlin commented:
@Rugvip, this is ready to be merged. Also, it would be good to include this in the release notes for `1.18.0`, how do we make sure that happens?

Rugvip commented:
@awanlin nice! 🎉

It'll be in the release notes, I've added a section to our internal draft of the notes

awanlin commented:
Thanks @Rugvip, and many, many, many thanks to @sanderaernouts for the work done in this PR! 🚀

sanderaernouts commented:
Thanks, @awanlin and @Rugvip, for your time and feedback 🚀 🎉
